Add type option to Worker constructor

aduh95 opened the pull request:
This PR adds a `type` option to the `Worker` constructor, inspired from [WorkerType in HTML spec](https://html.spec.whatwg.org/multipage/workers.html#workertype), and support for `data:` URLs. 

* When `eval` is true OR `filename` is a `data:` URL object, the `type` option explicitly tells Node.js what kind of JS to expect (ESM or CJS).
* Otherwise, the option is ignored and the usual file extension rules is used.

```js
// use ESM loader to load the worker.js file
new Worker('./worker.js', { type: 'module' });

// eval ESM
new Worker('export default import.meta.url', { eval: true, type: 'module' });

// data: URL
new Worker(
  new URL('data:text/javascript,export default import.meta.url'),
  { type: 'module' }
);
```

~Note: `new Worker('./worker.cjs', {type: 'module'})` would load `worker.cjs` as ES module and `new Worker('./worker.mjs', {type: 'classic'})` would load `worker.mjs` as CJS. I don't expect anyone to have a use-case for this edge-case, but I wanted to clarify it's a possibility with this PR's implementation.~ _This has been reverted based on this PR's comments._

I had to tweak some internal functions, I have put those changes in separate commits to make reviewing them easier.

Fixes: #30682

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

aduh95 commented:
Fixes #30682 btw

MylesBorins commented:
This seems reasonable to me. I've added to the modules agenda for visibility

guybedford commented:
It would help to understand a driving use case for needing this feature.

We don't have a problem with the fact that extensionless files cannot be imported from ES modules, so I'm just wondering why workers should be different.

guybedford commented:
Also I'm wondering if worker evaluation might be achieved with data URIs?

aduh95 commented:
> We don't have a problem with the fact that extensionless files cannot be imported from ES modules, so I'm just wondering why workers should be different.

Maybe should split this PR so we can discuss the two issues separately? My thinking was to make Node.js worker implementation closer to the HTML spec when using `type` option, and give users a way to shortcut the extension checking.

aduh95 commented:
After digging a bit more, it turns out the implementation is sloppier than I first imagined. I'll be waiting on modules team feedback if we need/want the actual feature before putting more work into it.

If we want to move this forward, I think I could benefit a lot from the changes on #31229, so maybe wait for it to land, then implement the feature requested on #21667, which could be used for a cleaner implementation.

jkrems commented:
> Note: new Worker('./worker.cjs', {type: 'module'}) would load worker.cjs as ES module and new Worker('./worker.mjs', {type: 'classic'}) would load worker.mjs as CJS.

That sounds like this circumvents the type detection entirely which isn't compatible with what the web equivalent does. Think of the file extension as our de-facto content-type header. On the web, `type: 'module'` wouldn't run a resource with [`content-type: application/node`](https://www.iana.org/assignments/media-types/application/node) as a JavaScript module.

I think the type option still makes sense to chose the module system (`require` by default, `import` when `type` is set to module). But I don't think it should affect content type resolution.

bmeck commented:
We could use a type assertion instead, which seems to be the difference of `as` (which asserts the type of a target from link tag / module attributes proposal) and `type` (which changes the loading system of a target).

However, per things like https://github.com/nodejs/node/pull/31388 which was brought about by extension-less support; we probably want to think about how this would expand in the future for formats outside of JS itself. I think if people were uncomfortable with `type: 'wasm'` in `package.json` it likely would be argument against using a `type` field in the constructor here as well.

Since the ESM loader does support `data:`, and `"bin":{"foo":"bin/foo.cjs"}` within `package.json` can point to a file with an extension it might also just be prudent to not do anything except having a switch to select the loader used for the initial worker endpoint; this is similar to how `shouldUseESMLoader` works https://github.com/nodejs/node/blob/f5ef7cd22231987a73af9f02cc5f6c80b6dbbd9b/lib/internal/modules/run_main.js#L23 in a dynamic manner and wouldn't appear to cause any new forms of conflicts where resources can be interpreted in multiple different ways.

addaleax commented:
@aduh95 This looks like it needs to be rebased.

@nodejs/modules-active-members Should this go ahead? It seems good to me.

bmeck commented:
We likely should not land this yet. Discussion of expanding to handle the non-JS cases isn't ironed out.

jkrems commented:
Should this option be called `inputType` since it's roughly equivalent to `--input-type` on the command line?

aduh95 commented:
> Should this option be called `inputType` since it's roughly equivalent to `--input-type` on the command line?

Should it be restricted to `eval: true` like the CLI option?

aduh95 commented:
I have removed the loader selection for this PR to unblock it.

I have also added support for `data:` URLs as https://github.com/nodejs/node/pull/31760#issuecomment-586323979 suggested.

aduh95 commented:
Closing this in favor of #34584.
